File: hw3/newton.py
# $MYHPSC/hw3/newton.py

def fvals_sqrt(x):
    """
    Returns the values of f(x) = x**2 - a
    and its derivative for any input x
    """
    fx  = x**2 - 4.
    fpx = 2.*x
    return fx, fpx

def solve(fvals_sqrt, x0, debug=False):
    """
    Iteratively solves for the root of a function 
    using newton's method
    Input parameters:
        fvals_sqrt - the function fvals_sqrt defined above
        x0 - inital guess
        debug_solve - a debug flag whether to print intermediate values
    """
    maxiter = 20
    fx, fpx = fvals_sqrt(x0) 
    if(debug == True):
        print "Initial guess: x = %22.15e" %(x0)
    xk = x0 - fx/fpx #testing the initial guess
    for i in range(maxiter):
        xk = xk - fx/fpx
        if(debug == True and i>0):
            print "After %i iterations, x = %22.15e" % (i, xk)
        fx,fpx = fvals_sqrt(xk)
        if(abs(fx) < 1e-14):
            break

    if(i < maxiter-1):
        return xk, i
    elif(i == maxiter-1 and abs(fx) < 1e-14):
        return
    elif(i == maxiter-1 and abs(fx) >= 1e-14):
        print "Could not converge to the solution within 20 iterations. Please try another initial guess"
        return xk,i
    else:
        print "solve exceeded 20 iterations. This should not have happened"
    
def test1(debug_solve=False):
    """
    Test Newton iteration for the square root with different initial
    conditions.
    """
    from numpy import sqrt
    for x0 in [1., 2., 100.]:
        print " "  # blank line
        x,iters = solve(fvals_sqrt, x0, debug=debug_solve)
        print "solve returns x = %22.15e after %i iterations " % (x,iters)
        fx,fpx = fvals_sqrt(x)
        print "the value of f(x) is %22.15e" % fx
        assert abs(x-2.) < 1e-14, "*** Unexpected result: x = %22.15e"  % x

File: lectures/w4-l10/newton/newton.f90
! $MYHPSC/w4-l10/newton/newton.f90

module newton

  !module parameters
  implicit none
  integer, parameter :: maxiter = 20
  real(kind=8), parameter :: tol = 1.d-14

contains

  subroutine solve(f, fp, x0, x, iters, debug)

    ! Estimate the zero of f(x) using Newton's method
    ! Input:
    !   f: the function of which we will find the root
    !   fp: function returnin the derivative f'
    !   x0: the intial guess
    !   debug: a logical, prints the number of iterations if debug is true
    ! Returns:
    !   x: the estimate of x satisfying f(x) = 0 (assumes Newton convergence)
    !   iters: the number of iterations required to converge 

    implicit none
    real(kind=8), intent(in) :: x0 ! cannot change, therefore we x = x0
    real(kind=8), external :: f, fp ! we are passing f and fp as functions defined elsewhere
    ! they return real(kind=8)
    logical, intent(in) :: debug
    real(kind=8), intent(out) :: x
    integer, intent(out) :: iters

    !Declare any local variables
    real(kind=8) :: deltax, fx, fxprime
    integer :: k

    ! initial guess
    x = x0

    if(debug) then
      print 11, x
11    format('Initial guess: x = ', e20.15) ! scientific notation, 20 total digits, 15
    endif                             ! 15 to the right of the decimal

  ! Newton iteration to find a zero of f(x)

  do k=1, maxiter

    ! evaluate function and its derivative
    fx = f(x)
    fxprime = fp(x)

    ! convergence test
    if (abs(fx) < tol) then ! should be approaching 0 if we are finding the root of the fxn
      exit ! jump out of do loop
    endif

    ! compute Newton increment x:
    deltax = fx/fxprime

    ! update x
    x = x - deltax

    if(debug) then
      print 12, k,x
      12 format('After' i3, ' iterations, x = ' e20.15)
    endif

   enddo


   if(k > maxiter) then
     ! might not have converged

     fx = f(x)
     if(abs(fx) > tol) then
       print *, "*** Warning: has not yet converged"
     endif
   endif

   ! number of iterations taken:
   iters = k-1

end subroutine solve

end module newton
